src/libs/costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -41,65 +41,76 @@
 #include <string>
 #include <algorithm>
 #include <vector>
-#include <tf2/convert.h>
-#include <tf2/utils.h>
 #include <tf2_geometry_msgs/tf2_geometry_msgs.h>
 
 using namespace std;
 
 namespace costmap_2d
 {
 
-void move_parameter(ros::NodeHandle& old_h, ros::NodeHandle& new_h, std::string name, bool should_delete = true)
+void move_parameter(rclcpp::Node::SharedPtr old_h, rclcpp::Node::SharedPtr new_h, std::string name,
+    bool should_delete = true)
 {
-  if (!old_h.hasParam(name))
+  auto parameters_client_old = std::make_shared<rclcpp::SyncParametersClient>(old_h);
+  auto parameters_client_new = std::make_shared<rclcpp::SyncParametersClient>(new_h);
+
+  if (!parameters_client_old->has_parameter(name)) {
     return;
+  }
 
-  XmlRpc::XmlRpcValue value;
-  old_h.getParam(name, value);
-  new_h.setParam(name, value);
-  if (should_delete) old_h.deleteParam(name);
+  std::vector<rclcpp::Parameter> param;
+  param = parameters_client_old->get_parameters({name});
+  parameters_client_new->set_parameters(param);
+  //TODO(bpwilcox): port delete parameter from client?
+  //if (should_delete) old_h.deleteParam(name);
 }
 
-Costmap2DROS::Costmap2DROS(const std::string& name, tf2_ros::Buffer& tf) :
-    layered_costmap_(NULL),
-    name_(name),
-    tf_(tf),
-    transform_tolerance_(0.3),
-    map_update_thread_shutdown_(false),
-    stop_updates_(false),
-    initialized_(true),
-    stopped_(false),
-    robot_stopped_(false),
-    map_update_thread_(NULL),
-    last_publish_(0),
-    plugin_loader_("costmap_2d", "costmap_2d::Layer"),
-    publisher_(NULL),
-    dsrv_(NULL),
-    footprint_padding_(0.0)
+Costmap2DROS::Costmap2DROS(const std::string & name, tf2_ros::Buffer & tf)
+  : layered_costmap_(NULL),
+  name_(name),
+  tf_(tf),
+  transform_tolerance_(0.3),
+  map_update_thread_shutdown_(false),
+  stop_updates_(false),
+  initialized_(true),
+  stopped_(false),
+  robot_stopped_(false),
+  map_update_thread_(NULL),
+  last_publish_(0),
+  plugin_loader_("costmap_2d", "costmap_2d::Layer"),
+  publisher_(NULL),
+  publish_cycle_(1),
+  footprint_padding_(0.0)
 {
-  // Initialize old pose with something
   tf2::toMsg(tf2::Transform::getIdentity(), old_pose_.pose);
 
-  ros::NodeHandle private_nh("~/" + name);
-  ros::NodeHandle g_nh;
+  auto private_nh = rclcpp::Node::make_shared(name_);
+  rclcpp::Node::SharedPtr g_nh;
 
   // get two frames
-  private_nh.param("global_frame", global_frame_, std::string("map"));
-  private_nh.param("robot_base_frame", robot_base_frame_, std::string("base_link"));
-
-  ros::Time last_error = ros::Time::now();
+  auto private_parameters_client = std::make_shared<rclcpp::SyncParametersClient>(private_nh);
+  global_frame_ =
+      private_parameters_client->get_parameter<std::string>("global_frame", std::string("map"));
+  robot_base_frame_ =
+      private_parameters_client->get_parameter<std::string>("robot_base_frame", std::string(
+        "base_link"));
+
+  auto clock = new rclcpp::Clock();
```

Comment:
This is a 'sort of' memory leak. You never delete this pointer. Just make it a local variable
```c++
rclcpp::Clock clock;
rclcpp::Time last_error = clock.now();
```